tools/stats/rpc.go

```diff
@@ -46,19 +46,74 @@ func getAPI(path string) (string, http.Header, error) {
 }
 
 func WaitForSyncComplete(ctx context.Context, napi api.FullNode) error {
+sync_complete:
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-time.After(3 * time.Second):
+		case <-time.After(5 * time.Second):
+			state, err := napi.SyncState(ctx)
+			if err != nil {
+				return err
+			}
+
+			for i, w := range state.ActiveSyncs {
+				if w.Target == nil {
+					continue
+				}
+
+				if w.Stage == api.StageSyncErrored {
+					log.Errorw(
+						"Syncing",
+						"worker", i,
+						"base", w.Base.Key(),
+						"target", w.Target.Key(),
+						"target_height", w.Target.Height(),
+						"height", w.Height,
+						"error", w.Message,
+						"stage", chain.SyncStageString(w.Stage),
+					)
+				} else {
+					log.Infow(
+						"Syncing",
+						"worker", i,
+						"base", w.Base.Key(),
+						"target", w.Target.Key(),
+						"target_height", w.Target.Height(),
+						"height", w.Height,
+						"stage", chain.SyncStageString(w.Stage),
+					)
+				}
+
+				if w.Stage == api.StageSyncComplete {
+					break sync_complete
+				}
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(5 * time.Second):
 			head, err := napi.ChainHead(ctx)
 			if err != nil {
 				return err
 			}
 
-			log.Printf("Height %d", head.Height())
+			timestampDelta := time.Now().Unix() - int64(head.MinTimestamp())
+
+			log.Infow(
+				"Waiting for reasonable head height",
+				"height", head.Height(),
+				"timestamp_delta", timestampDelta,
+			)
 
-			if time.Now().Unix()-int64(head.MinTimestamp()) < build.BlockDelay {
+			// If we get within 20 blocks of the current exected block height we
+			// consider sync complete. Block propagation is not always great but we still
+			// want to be recording stats as soon as we can
+			if timestampDelta < build.BlockDelay*20 {
```

Comment:
To further explain this: If we are experience any sort of delayed block propagation and we restart `stats` we are stuck waiting for the network to settle down.

Reply:
Might need to remove this all together if it won't cause problems. When we have issues such as we currently are and the chain is far behind the expected block height we can't gather stats.